sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1199,13 +1207,189 @@ def task_run(self, sfp_error_event, y_cable_presence):
             return
 
         self.task_process = multiprocessing.Process(target=self.task_worker, args=(
-            self.task_stopping_event, sfp_error_event, y_cable_presence))
+            self.task_stopping_event, sfp_error_event, y_cable_presence, self.event_queue))
         self.task_process.start()
 
     def task_stop(self):
         self.task_stopping_event.set()
         os.kill(self.task_process.pid, signal.SIGKILL)
 
+    def notify_port_change_event(self, port_change_event):
+        """Called by main thread. When main thread detects a port configuration add/remove, it puts an event to 
+           the event queue
+
+        Args:
+            port_change_event (object): port change event
+        """
+        self.event_queue.put_nowait(port_change_event)
+
+    def handle_port_change_event(self, event_queue, stopping_event, y_cable_presence):
+        """Handle port change event until the queue is empty, update local port mapping and DB data accordingly. 
+        """
+        while True:
+            try:
+                port_change_event = event_queue.get_nowait()
+                if port_change_event.event_type == PortChangeEvent.PORT_REMOVE:
+                    self.on_remove_logical_port(port_change_event)
+                    # Update y_cable related database once a logical port is removed
+                    y_cable_helper.change_ports_status_for_y_cable_change_event(
+                        {port_change_event.port_name:sfp_status_helper.SFP_STATUS_REMOVED}, 
+                        self.port_mapping, 
+                        y_cable_presence, 
+                        stopping_event)
+                    self.port_mapping.handle_port_change_event(port_change_event)
+                elif port_change_event.event_type == PortChangeEvent.PORT_ADD:
+                    self.port_mapping.handle_port_change_event(port_change_event)
```

Comment:
No, because the call order matters. When the event is remove, `y_cable_helper.change_ports_status_for_y_cable_change_event` takes port_mapping as an argument, the removing port must be in port_mapping at that time or it would cause issues. See code in the function:

```python
            # Get the asic to which this port belongs
            asic_index = port_mapping.get_asic_id_for_logical_port(logical_port_name)
            if asic_index is None:
                helper_logger.log_warning("Got invalid asic index for {}, ignored".format(logical_port_name))
                continue
```